text/0004-api-history-schema.md

```diff
@@ -0,0 +1,249 @@
+which will include historical information about changes to the various APIs in Electron.
+
+## Motivation
+
+The API history blocks need to be written in a well-designed format in order to
+convey information efficiently and ensure readability. These blocks will also be
+transformed into tables in `electron/website` to allow users to see what/when/how
+Electron APIs have changed. In order to do so, the blocks need to have a consistent
+and parsable structure.
+
+## Guide-level explanation
+
+Before this RFC, when a developer created a new feature they added documentation
+for it like this:
+
+`electron/docs/api/browser-window.md`
+
+```markdown
+#### `win.setTrafficLightPosition(position)` _macOS_
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with `titleBarStyle` set to `hidden`.
+```
+
+After this RFC, they will consult the schema, and also add an API history block:
+
+````markdown
+#### `win.setTrafficLightPosition(position)` _macOS_
+
+<!--
+```YAML history
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+```
+-->
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with
+`titleBarStyle` set to `hidden`.
+````
+
+Which, after several PRs, may reach this state:
+
+```yaml
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+changes:
+  - pr-url: https://github.com/electron/electron/pull/26789
+    description: Made `trafficLightPosition` option work for `customButtonOnHover` window.
+deprecated:
+  - pr-url: https://github.com/electron/electron/pull/37094
+    breaking-changes-header: deprecated-browserwindowsettrafficlightpositionposition
+```
+
+> [!NOTE]
+>
+> - *"Why no version numbers?"*
+>   - They will be derived from the PRs. This removes the need to change API History
+> on backports.
+>
+> - *"Why no `removed` property?"*
+>   - When an API is removed from Electron, it is also removed from the documentation.
+>
+> For further details, see [**Rationale and alternatives**](#rationale-and-alternatives).
+
+Meanwhile, the Electron website displays this information nicely in a table.
+Users can then use this information to help them upgrade their Electron version,
+migrate from deprecated APIs, etc.
+
+The schema provides information to the developer about what to include in the
+block, the minimum/maximum length of text, the required properties, etc.
+
+Documentation for older APIs will also eventually be changed to include these
+API history blocks.
+
+## Reference-level explanation
+
+The JSON Schema looks like this:
+
+```json
+{
+  "title": "JSON schema for API history blocks in Electron documentation",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$comment": "If you change this schema, remember to edit the TypeScript interfaces in the linting script.",
+  "definitions": {
```

Comment:
@dsanders11 How do you feel about adding a version property?

```suggestion
  "$comment": "If you change this schema, remember to edit the TypeScript interfaces in the linting script.",
  "version": "1.0.0",
  "definitions": {
```

Reply:
Not sure it adds much in practice, and we'll probably forget to update it. 😅